### PR TITLE
Avoid duplications in RISC-V exception handlers

### LINKED ISSUE #61: Avoid duplications in RISC-V exception handlers
Code duplication in the function body of the RISC-V exception handlers might be a maintenance headache. That is, we shall refine these functions in `src/emulate.c`:
* `rv_except_insn_misaligned`
* `rv_except_load_misaligned`
* `rv_except_store_misaligned`
* `rv_except_illegal_insn`
* `rv_except_breakpoint` proposed in #60

Code generation using preprocessor macros may be an approach to avoid such duplications.

### PR DESCRIPTION
**Pull request by Risheng1128**
Use preprocessor to avoid duplications in exception handlers, 
and create a list to define the RISC-V exception code.

Resolve #61

### CODE REVIEW

File: src/riscv.h [hunk 1]
```diff
@@ -42,6 +42,17 @@ enum {
     rv_reg_t6,
 };
 
+/* RISC-V exception code list */
+enum {
```
Comment (jserv) [inline src/riscv.h]: Move the enumeration to C source.

Comment (jserv) [reply src/riscv.h]: With the [X macro](https://en.wikipedia.org/wiki/X_Macro) trick, we can write down shorter code.
Check https://github.com/sysprog21/rv32emu/blob/master/src/syscall.c (SUPPORTED_SYSCALLS)

**Comment by Risheng1128** [comment]
Move the enumeration into source file.
Use [X macro](https://en.wikipedia.org/wiki/X_Macro) trick to construct the list.

File: src/emulate.c [hunk 2]
```diff
@@ -24,160 +24,57 @@ extern struct target_ops gdbstub_ops;
...
-    switch (mode) {
-    case 0: /* DIRECT: All exceptions set PC to base */
-        rv->PC = base;
-        break;
-    case 1: /* VECTORED: Asynchronous interrupts set PC to base + 4 * code */
-        rv->PC = base + 4 * code;
-        break;
+/* RISC-V exception code list */
+#define GET_EXCEPTION_CODE(type) rv_exception_code_##type
+#define __EXCEPTION_LIST                                        \
```
Comment (jserv) [inline src/emulate.c]: Rename to `RV_EXCEPTION_LIST`.

File: src/emulate.c [hunk 3]
```diff
@@ -24,160 +24,57 @@ extern struct target_ops gdbstub_ops;
...
+    _(insn_misaligned, 0)  /* Instruction address misaligned */ \
+    _(insn_fault, 1)       /* Instruction access fault */       \
+    _(illegal_insn, 2)     /* Illegal instruction */            \
+    _(breakpoint, 3)       /* Breakpoint */                     \
+    _(load_misaligned, 4)  /* Load address misaligned */        \
+    _(load_fault, 5)       /* Load access fault */              \
+    _(store_misaligned, 6) /* Store/AMO address misaligned */
+
+enum {
+#define _(type, code) GET_EXCEPTION_CODE(type) = code,
```
Comment (jserv) [inline src/emulate.c]: Using `enum`, it is not necessary to specify the assignment.

File: src/emulate.c [hunk 4]
```diff
@@ -24,160 +24,57 @@ extern struct target_ops gdbstub_ops;
...
-        break;
-    case 1: /* VECTORED: Asynchronous interrupts set PC to base + 4 * code */
-        rv->PC = base + 4 * code;
-        break;
-    }
-
-    /* mcause (Machine Cause Register): store exception code */
-    rv->csr_mcause = code;
-}
+EXCEPTION_HANDLER_IMPL(insn_misaligned)
```
Comment (jserv) [inline src/emulate.c]: Rewrite as following:
```c
/* RISC-V exception handlers */
#define _(type) EXCEPTION_HANDLER_IMPL(type)
RV_EXCEPTION_LIST
#undef _
```

**Review by jserv — CHANGES_REQUESTED** [review]
Append `Close #61` at the end of git commit message.

**Comment by Risheng1128** [comment]
Becaue there are some exception handlers unused like `insn_fault` and 
`load_fault`, I add the function attributes `unused` to prevent compiler
warning.

**Comment by jserv** [comment]
Thank @Risheng1128 for contributing!

**Pull request merged**